### DISCUSSION TITLE
使用实时功能，推送文档到kafka后不生效

### DISCUSSION THREAD

**Discussion by duocairen** [alibaba/havenask#89] [Q&A] [answered]
按照  [实时功能使用文档](https://github.com/alibaba/havenask/wiki/%E5%AE%9E%E6%97%B6%E5%8A%9F%E8%83%BD%E4%BD%BF%E7%94%A8%E6%96%87%E6%A1%A3)启动了example里边realtime的case并且配置了kafka集群地址；
按照 [实时推送文档格式](https://github.com/alibaba/havenask/wiki/%E6%96%87%E6%A1%A3%E6%A0%BC%E5%BC%8F)往kafka推送了文档，但是通过命令行执行select * from in0查表数据没变化，看不到新插入的数据；

kafka的分区数设置为1，因此应该不存在分区hash不对的问题，推送到kafka的方式是通过spring kafkaTemplate推送
@RequestMapping({"/send"})
    public String send(String title, int id) {
        String data = "CMD=add^_\n"
                + "title=" + title + "^_\n"
                + "id=" + id + "^_\n"
                + "hits=" + RandomUtil.randomInt(1, 100) + "^_\n"
                + "createtime=" + (DateTime.now().getMillis() / 1000) + "^_\n"
                + "^^\n";
        log.info("发送文档:{}", data);
        kafkaTemplate.send("ha3_test", String.valueOf(id), data);
        return "success";
    }

**Comment by WenyiDuan** [comment]
先看下ha3.log中是否打印了日志"kafka reader init succuss"，确保reader正常初始化了

**Reply by duocairen** [reply]
看ha3.log有正常打印"kafka reader init succuss"，已经排查到了，是java里特殊字符需要使用编码形式

**Comment by weizijun** [comment, marked as answer]
java里特殊字符需要这样传递：
行末：message.append("CMD=add\u001F\n");
结尾：message.append("\u001E\n");

**Reply by duocairen** [reply]
改成这样的编码形式可以了

**Comment by WenyiDuan** [comment]
发送消息的时候需要指定消息的key，否则校验的时候key为null会被过滤掉

**Reply by duocairen** [reply]
在发送时有使用id作为消息key，排查了是编码问题
